spec/castle_devise/integration/password_update_spec.rb

```diff
@@ -0,0 +1,167 @@
+  end
+
+  context "when profile update hooks are disabled" do
+    around do |example|
+      User.castle_hooks[:profile_update] = false
+      example.run
+      User.castle_hooks[:profile_update] = true
+    end
+
+    before do
+      allow(facade).to receive(:risk).and_return(login_castle_risk_response)
+
+      send_password_update
+    end
+
+    it "does not use risk action for the profile_update event" do
+      expect(facade).not_to have_received(:risk).with(
+        event: "$profile_update",
+        status: "$attempted",
+        context: have_attributes(email: email, resource: user)
+      )
+    end
+
+    it "does not log the profile_update event" do
+      expect(facade).not_to have_received(:log)
+    end
+
+    it "updates the password" do
+      expect(user.reload.valid_password?(new_password)).to eq(true)
+    end
+  end
+
+  context "when profile update hooks are enabled" do
+    before do
+      allow(facade).to receive(:risk).and_return(
+        login_castle_risk_response, profile_update_castle_risk_response
+      )
+
+      send_password_update
+    end
+
+    context "when password successfully changed" do
+      it "updates the password" do
+        expect(user.reload.valid_password?(new_password)).to eq(true)
+      end
+
+      it "performs risk action with profile_update event" do
+        expect(facade).to have_received(:risk).with(
+          event: "$profile_update",
+          status: "$attempted",
+          context: have_attributes(email: email, resource: user)
+        )
+      end
+
+      it "logs profile_update event with succeeded status" do
+        expect(facade).to have_received(:log).with(
+          event: "$profile_update",
+          status: "$succeeded",
+          context: have_attributes(email: email, resource: user)
+        )
+      end
+    end
+
+    context "when password failed to change" do
+      let(:current_password) { "abcdef" }
+
+      it "does not update the password" do
+        expect(user.reload.valid_password?(password)).to eq(true)
+      end
+
+      it "performs risk action with profile_update event" do
+        expect(facade).to have_received(:risk).with(
+          event: "$profile_update",
+          status: "$attempted",
+          context: have_attributes(email: email, resource: user)
+        )
+      end
+
+      it "logs profile_update event with failed status" do
+        expect(facade).to have_received(:log).with(
+          event: "$profile_update",
+          status: "$failed",
+          context: have_attributes(email: email, resource: user)
+        )
+      end
+    end
+  end
+
+  context "when profile update hooks are enabled and there are errors" do
+    context "when Castle raises InvalidParametersError" do
+      before do
+        call_risk_count = 0
+        allow(facade).to receive(:risk) do
+          call_risk_count += 1
+          call_risk_count >= 2 ? raise(Castle::InvalidParametersError) : login_castle_risk_response
```

Comment:
Why do we raise after the 2nd call? Don't we do only one call to `risk` in the code?

Reply:
We perform request to `Risk` during signing in and then we don't want to raise any errors. When performing the second call we want because that's our case here for `$profile_update`.